genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/JobSetupServiceImpl.java

```diff
@@ -0,0 +1,497 @@
+        try {
+            Files.createDirectory(jobDirectory.toPath());
+        } catch (final IOException e) {
+            throw new SetUpJobException("Failed to create job directory", e);
+        }
+
+        // Get DTOs
+        final List<JobSpecification.ExecutionResource> applications = jobSpecification.getApplications();
+        final JobSpecification.ExecutionResource cluster = jobSpecification.getCluster();
+        final JobSpecification.ExecutionResource command = jobSpecification.getCommand();
+
+        // Make a list of all entity dirs
+        final List<Path> entityDirectories = Lists.newArrayList(
+            PathUtils.jobClusterDirectoryPath(jobDirectory, cluster.getId()),
+            PathUtils.jobCommandDirectoryPath(jobDirectory, command.getId())
+        );
+
+        applications.stream()
+            .map(JobSpecification.ExecutionResource::getId)
+            .map(appId -> PathUtils.jobApplicationDirectoryPath(jobDirectory, appId))
+            .forEach(entityDirectories::add);
+
+        // Make a list of directories to create
+        // (only "leaf" paths, since createDirectories creates missing intermediate dirs
+        final List<Path> directoriesToCreate = Lists.newArrayList();
+
+        // Add config and dependencies for each entity
+        entityDirectories.forEach(
+            entityDirectory -> {
+                directoriesToCreate.add(PathUtils.jobEntityDependenciesPath(entityDirectory));
+                directoriesToCreate.add(PathUtils.jobEntityConfigPath(entityDirectory));
+            }
+        );
+
+        // Add logs dir
+        directoriesToCreate.add(PathUtils.jobGenieLogsDirectoryPath(jobDirectory));
+
+        // Create directories
+        for (final Path path : directoriesToCreate) {
+            try {
+                Files.createDirectories(path);
+            } catch (final Exception e) {
+                throw new SetUpJobException("Failed to create directory: " + path, e);
+            }
+        }
+
+        return jobDirectory;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<File> downloadJobResources(
+        final JobSpecification jobSpecification,
+        final File jobDirectory
+    ) throws SetUpJobException {
+
+        final List<java.io.File> setupFiles = Lists.newArrayList();
+
+        // Create download manifest for dependencies, configs, setup files for cluster, applications, command, job
+        final DownloadService.Manifest jobDownloadsManifest =
+            createDownloadManifest(jobDirectory, jobSpecification, setupFiles);
+
+        // Download all files into place
+        try {
+            this.downloadService.download(jobDownloadsManifest);
+        } catch (final DownloadException e) {
+            throw new SetUpJobException("Failed to download job dependencies", e);
+        }
+
+        return setupFiles;
+    }
+
+    private DownloadService.Manifest createDownloadManifest(
```

Comment:
This is a relatively large method that might be better served for readability to break it up? More of a small thing than a need to do or anything like that